std/machines/hash/keccakf16.asm

```diff
@@ -0,0 +1,608 @@
+    //     /// The bits of `A''[0, 0]`.
+    //     pub a_prime_prime_0_0_bits: [T; 64],
+
+    //     /// ```ignore
+    //     /// A'''[0, 0, z] = A''[0, 0, z] ^ RC[k, z]
+    //     /// ```
+    //     pub a_prime_prime_prime_0_0_limbs: [T; U64_LIMBS],
+    // }
+
+    pol commit preimage[5 * 5 * 4];
+    pol commit a[5 * 5 * 4];
+    pol commit c[5 * 64];
+    pol commit c_prime[5 * 64];
+    pol commit a_prime[5 * 5 * 64];
+    pol commit a_prime_prime[5 * 5 * 4];
+    pol commit a_prime_prime_0_0_bits[64];
+    pol commit a_prime_prime_prime_0_0_limbs[4];
+
+    // Initially, the first step flag should be 1 while the others should be 0.
+    // builder.when_first_row().assert_one(local.step_flags[0]);
+    // for i in 1..NUM_ROUNDS {
+    //     builder.when_first_row().assert_zero(local.step_flags[i]);
+    // }
+    // for i in 0..NUM_ROUNDS {
+    //     let current_round_flag = local.step_flags[i];
+    //     let next_round_flag = next.step_flags[(i + 1) % NUM_ROUNDS];
+    //     builder
+    //         .when_transition()
+    //         .assert_eq(next_round_flag, current_round_flag);
+    // }
+
+    let step_flags: col[NUM_ROUNDS] = array::new(NUM_ROUNDS, |i| |row| if row % NUM_ROUNDS == i { 1 } else { 0 } );
+
+    // let main = builder.main();
+    // let (local, next) = (main.row_slice(0), main.row_slice(1));
+    // let local: &KeccakCols<AB::Var> = (*local).borrow();
+    // let next: &KeccakCols<AB::Var> = (*next).borrow();
+
+    // let first_step = local.step_flags[0];
+    // let final_step = local.step_flags[NUM_ROUNDS - 1];
+    // let not_final_step = AB::Expr::one() - final_step;
+
+    let first_step: expr = step_flags[0]; // aliasing instead of defining a new fixed column
+    let final_step: expr = step_flags[NUM_ROUNDS - 1];
+    // Equals to 1 for all non-last rows in all 24-row blocks. Forced to zero for the very last row of the submachine (if partial block).
+    // This is used to deactivate constraints that reference the next row, whenever we are at the latch row or the last row of the trace (so that we don't incorrectly cycle to the first row).
+    col fixed internal_row = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0]* + [0];
+
+    // // If this is the first step, the input A must match the preimage.
+    // for y in 0..5 {
+    //     for x in 0..5 {
+    //         for limb in 0..U64_LIMBS {
+    //             builder
+    //                 .when(first_step)
+    //                 .assert_eq(local.preimage[y][x][limb], local.a[y][x][limb]);
+    //         }
+    //     }
+    // }
+
+    array::zip(preimage, a, |p_i, a_i| first_step * (p_i - a_i) = 0);
+
+    // // The export flag must be 0 or 1.
+    // builder.assert_bool(local.export);
+
+    // force_bool(export);
+
+    // // If this is not the final step, the export flag must be off.
+    // builder
+    //     .when(not_final_step.clone())
+    //     .assert_zero(local.export);
+
+    // not_final_step * export = 0;
```

Comment:
The export flag is really just the latch column `final_step`. Added a comment.

Reply:
FYI here's the definition of the `export` flag in Plonky3:

```
    /// The `i`th value is set to 1 if we are in the `i`th round, otherwise 0.
    pub step_flags: [T; NUM_ROUNDS],

    /// A register which indicates if a row should be exported, i.e. included in a multiset equality
    /// argument. Should be 1 only for certain rows which are final steps, i.e. with
    /// `step_flags[23] = 1`.
    pub export: T,
```

Reply:
To be honest I'm still a bit confused by the use of the selector column(s) in permutations. Are they submachine inputs from the main machine? Are they also needed in lookups?

Reply:
They are not needed in lookups.

If a machine is linked via a lookup argument, it would look like this:
```
main.instr_foo $ [ <operation_id>, arg1, arg2, ... ]
in
submachine.LATCH $ [ submachine.operation_id, submachine.arg1, submachine.arg2 ]
```
(In your case, the "latch" is the `final_step` column.)

The lookup is a subset argument, all tuples on the LHS (those rows where `main.instr_foo == 1`) must be in the set defined by the RHS (those rows where `submachine.LATCH == 1`. One way to satisfy this is to create a block for each call of the main machine and then fill the remaining rows with copies of the first block.

The permutation argument is stricter: You can't have anything on the RHS that does not appear in the LHS. To still satisfy the constraint, we need to give the prover the flexibility to indicate which "latch rows" belong to actual blocks, and which are just padded blocks.

For this, we introduce a new selector witness column, constrained to be 0 or 1. The constraint is then:
```
main.instr_foo $ [ <operation_id>, arg1, arg2, ... ]
is
(submachine.selector * submachine.LATCH) $ [ submachine.operation_id, submachine.arg1, submachine.arg2 ]
```

The term `(submachine.selector * submachine.LATCH)` can only evaluate to 1 if `submachine.LATCH == 1`, but the prover additionally has the ability to remove rows from the RHS by setting the selector to zero.

You'd need a new selector column for each permutation, which is why there is an array of selector columns.